### PR TITLE
Add custom date and object fields

### PR DESCRIPTION
**Pull request by jbalboni**
Related to #4608 

This adds a date field and adds some tweaks to the object field to allow us to specify titles and descriptions differently.

Changes are:
1. New default DateWidget with our current partial date validation (i.e. just the year is validated)
2. Copied over ObjectField from the library and made a couple tweaks: title and description are now pulled from the uiSchema instead of the regular schema. This should put almost all of the information we care about on the front end into the uiSchema
3. Updated FieldTemplate to also pull the title and description from uiSchema
4. Split widgets into their own files in a folder to make it clearer
5. Created a uiDefinitions file for common uiSchema configurations.

The main changes to review here are DateWidget and the changes to the form config. Almost all the other changes are copied from the library or code being moved.

screen shot 2017-01-09 at 4 06 54 pm
screen shot 2017-01-09 at 4 05 05 pm
screen shot 2017-01-09 at 4 04 59 pm

### CODE REVIEW

File: src/js/common/schemaform/ObjectField.jsx
Comment (jbalboni) [inline src/js/common/schemaform/ObjectField.jsx]: This block has changes from me

File: src/js/common/schemaform/ObjectField.jsx
Comment (jbalboni) [inline src/js/common/schemaform/ObjectField.jsx]: These two lines are the other changes from me.

File: src/js/common/schemaform/widgets/DateWidget.jsx
Comment (akainic) [inline src/js/common/schemaform/widgets/DateWidget.jsx]: Where did this come from?

**Review by akainic — COMMENTED** [review]
This looks great! Just a few questions.

File: src/js/common/schemaform/ObjectField.jsx
Comment (jbalboni) [inline src/js/common/schemaform/ObjectField.jsx]: I've already done two, but I haven't heard back in a few weeks. So, I would like to submit as much as we can, but I'm not optimistic about the maintainers being responsive.

File: src/js/common/schemaform/uiDefinitions.js
Comment (jbalboni) [inline src/js/common/schemaform/uiDefinitions.js]: To make it easier to share the schema with the backend, yes. I think the more stuff we put in uiSchema the better, since we can update it directly in this repo and don't have to worry about shared backend code.

File: src/js/common/schemaform/widgets/DateWidget.jsx
Comment (jbalboni) [inline src/js/common/schemaform/widgets/DateWidget.jsx]: This is the new date widget I wrote based on our existing date component.

File: src/js/common/schemaform/ObjectField.jsx
Comment (akainic) [inline src/js/common/schemaform/ObjectField.jsx]: That makes sense. For now can we document the changes we've made somewhere so we can keep track of those in case the first 2 PRs ever get merged?

**Review by akainic — APPROVED** [review]
LGTM! Feel free to merge when you're ready and when the routes are commented out again.

**Pull request merged**
